File: BehavioralPatterns/Command/Invoker.py
from Order import Order

class Waitress():
    def __init__(self):
        self.__snakeQty = 2
        self.__drinkQty = 4
        self.__orderList = []
    
    #接收訂單
    def setOrder(self, order: Order):
        if(order.name == "snakeOrder"):
            if (self.__snakeQty <= 0):
                print("Snake sold out")
            else:
                print("Add snake order")
                self.__snakeQty = self.__snakeQty - 1
                self.__orderList.append(order)

        if(order.name == "drinkOrder"):
            if (self.__drinkQty <= 0):
                print("Drink sold out")
            else:
                print("Add drink order")
                self.__drinkQty = self.__drinkQty - 1
                self.__orderList.append(order)

    #取消訂單
    def cancelOrder(self, order: Order):
        if (order.name == "snakeOrder"):
            self.__snakeQty = self.__snakeQty + 1
            print("Cancel snake order")

        if (order.name == "drinkOrder"):
            self.__drinkQty = self.__drinkQty + 1
            print("Cancel drink order")
            self.__orderList.remove(order)
    
    #將訂單送到廚房
    def notifyBaker(self): 
        for order in self.__orderList:
            order.sendOrder()
        self.__orderList.clear()

File: DESIGN-CSharp/Command/Invoker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Command
{
    //服務生
    public class Waitress
    {
        private int snakeQty = 2;

        private int drinkQty = 4;

        private List<Order> orderList = new List<Order>();

        //接收訂單
        public void setOrder(Order order) 
        {
            if (order.name.Equals("snakeOrder")) 
            {
                if (this.snakeQty <= 0)
                {
                    Console.WriteLine("Snake sold out");
                }
                else
                {
                    Console.WriteLine("Add snake order");
                    snakeQty--;
                    orderList.Add(order);
                }
            }

            if (order.name.Equals("drinkOrder"))
            {
                if (this.drinkQty <= 0)
                {
                    Console.WriteLine("Drink sold out");
                }
                else
                {
                    Console.WriteLine("Add drink order");
                    drinkQty--;
                    this.orderList.Add(order);
                }
            }
        }

        //取消訂單
        public void cancelOrder(Order order)
        {
            if (order.name.Equals("snakeOrder"))
            {
                snakeQty++;
                Console.WriteLine("Cancel snake order");
            }

            if (order.name.Equals("drinkOrder"))
            {
                drinkQty++;
                Console.WriteLine("Cancel drink order");
            }

            this.orderList.Remove(order);
        }

        //將訂單送到廚房
        public void notifyBaker() 
        {
            foreach (var order in this.orderList)
            {
                order.sendOrder();
            }

            this.orderList.Clear();
        }
    }
}
